README.md
# stactools-threedep

- Name: threedep
- Package: `stactools.threedep`
- PyPI: https://pypi.org/project/stactools-threedep/
- Owner: @gadomski
- Dataset homepage: https://www.usgs.gov/core-science-systems/ngp/3dep
- STAC extensions used:
  - [proj](https://github.com/stac-extensions/projection/)
  - [raster](https://github.com/stac-extensions/raster/)
  - [scientific](https://github.com/stac-extensions/scientific/)

This package creates STAC items for USGS 3DEP (formerly NED) elevation data.

## Examples

### STAC objects

- [Catalog](examples/catalog.json)
- [Collection](examples/usgs-3dep-1/collection.json)
- [Item](examples/usgs-3dep-1/n41w106-1/n41w106-1.json)

### Command-line usage

Create a catalog for all 3DEP data.
This catalog will contain two collections, one for the 1 arc-second data and one for the 1/3 arc-second data:

```bash
stac threedep create-catalog destination
```

Create a catalog for a single lat+lon id:

```bash
stac threedep create-catalog --asset-id n41w106 examples
```

Download all XML metadata:

```bash
stac threedep download-metadata destination
```

Fetch all available IDs:

```bash
stac threedep fetch-ids
```

Use `stac threedep --help` to see all subcommands and options.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project attempts to match the major and minor versions of [stactools](https://github.com/stac-utils/stactools) and increments the patch number as needed.

## [Unreleased]

### Added

- pre-commit, black, mypy, and more ([#11](https://github.com/stactools-packages/threedep/pull/11))
- Raster and scientific extensions, package description ([#12](https://github.com/stactools-packages/threedep/pull/12))
- Fix `asset_ids` definition in create-catalog CLI ([#14](https://github.com/stactools-packages/threedep/issues/14))

### Deprecated

### Removed

- Python 3.6 support ([#11](https://github.com/stactools-packages/threedep/pull/11)) 
- `threedep:region` ([#12](https://github.com/stactools-packages/threedep/pull/12))

### Fixed

- Proj extension, license, asset paths ([#12](https://github.com/stactools-packages/threedep/pull/12))
- GPKG path ([#13](https://github.com/stactools-packages/threedep/pull/13))

## v0.2.0

The first release as its own repository.


tests/test_stac.py
import datetime
import unittest

import pytest
from pystac.extensions.projection import ProjectionExtension
from pystac.extensions.raster import RasterExtension

from stactools.threedep import stac
from stactools.threedep.constants import USGS_FTP_BASE
from tests import test_data


class CreateItemTest(unittest.TestCase):
    def test_create_item_1(self) -> None:
        path = test_data.get_path(
            "data-files/base/1/TIFF/current/n41w106/USGS_1_n41w106.xml"
        )
        item = stac.create_item(path)
        self.assertEqual(item.id, "n41w106-1")
        self.assertTrue(item.geometry is not None)
        self.assertTrue(item.bbox is not None)
        self.assertEqual(
            item.datetime,
            datetime.datetime(2020, 10, 10, 0, 0, 0, tzinfo=datetime.timezone.utc),
        )
        self.assertEqual(
            item.common_metadata.start_datetime,
            datetime.datetime(2020, 10, 3, 0, 0, 0, tzinfo=datetime.timezone.utc),
        )
        self.assertEqual(
            item.common_metadata.end_datetime,
            datetime.datetime(2020, 10, 10, 0, 0, 0, tzinfo=datetime.timezone.utc),
        )

        data = item.assets["data"]
        self.assertEqual(
            data.href,
            (
                "https://prd-tnm.s3.amazonaws.com/StagedProducts"
                "/Elevation/1/TIFF/current/n41w106/USGS_1_n41w106.tif"
            ),
        )
        self.assertEqual(data.title, "USGS 1 Arc Second n41w106 20220331")
        self.assertIsNone(data.description)
        self.assertTrue(
            data.media_type, "image/tiff; application=geotiff; profile=cloud-optimized"
        )
        self.assertTrue(data.roles, ["data"])
        raster = RasterExtension.ext(data)
        assert raster.bands
        assert len(raster.bands) == 1
        self.assertDictEqual(
            raster.bands[0].to_dict(),
            {
                "nodata": -999999.0,
                "sampling": "point",
                "data_type": "float32",
                "spatial_resolution": 30,
            },
        )

        data = item.assets["metadata"]
        self.assertEqual(
            data.href,
            (
                "https://prd-tnm.s3.amazonaws.com/StagedProducts"
                "/Elevation/1/TIFF/current/n41w106/USGS_1_n41w106.xml"
            ),
        )
        self.assertTrue(data.title is None)
        self.assertTrue(data.description is None)
        self.assertEqual(data.media_type, "application/xml")
        self.assertEqual(data.roles, ["metadata"])

        data = item.assets["thumbnail"]
        self.assertEqual(
            data.href,
            (
                "https://prd-tnm.s3.amazonaws.com/StagedProducts/Elevation/"
                "1/TIFF/current/n41w106/USGS_1_n41w106.jpg"
            ),
        )
        self.assertTrue(data.title is None)
        self.assertTrue(data.description is None)
        self.assertEqual(data.media_type, "image/jpeg")
        self.assertEqual(data.roles, ["thumbnail"])

        data = item.assets["gpkg"]
        self.assertEqual(
            data.href,
            (
                "https://prd-tnm.s3.amazonaws.com/StagedProducts/Elevation/"
                "1/TIFF/current/n41w106/USGS_1_n41w106.gpkg"
            ),
        )
        self.assertTrue(data.title is None)
        self.assertEqual(
            data.description,
            "Spatially-referenced polygonal footprints of the source data used "
            "to assemble the DEM layer. The attributes of each source dataset, "
            "such as original spatial resolution, production method, and date "
            "entered into the standard DEM, are linked to these footprints.",
        )
        self.assertEqual(data.media_type, "application/geopackage+sqlite3")
        self.assertEqual(data.roles, ["metadata"])

        link = next(link for link in item.links if link.rel == "via")
        self.assertTrue(link is not None)
        self.assertEqual(
            link.target,
            (
                "https://prd-tnm.s3.amazonaws.com/StagedProducts"
                "/Elevation/1/TIFF/current/n41w106/USGS_1_n41w106.xml"
            ),
        )

        projection = ProjectionExtension.ext(item)
        self.assertEqual(projection.epsg, 4269)
        self.assertEqual(projection.shape, (3612, 3612))
        self.assertEqual(
            projection.transform,
            [
                0.00027777777803598015,
                0.0,
                -106.00166666708242,
                0.0,
                -0.000277777777786999,
                41.00166666678416,
            ],
        )

        item.validate()

    def test_create_item_1_weird_date(self) -> None:
        path = test_data.get_path("data-files/one-offs/USGS_1_n19w090.xml")
        item = stac.create_item(path)
        self.assertEqual(
            item.datetime,
            datetime.datetime(2013, 11, 1, 0, 0, 0, tzinfo=datetime.timezone.utc),
        )

    def test_create_item_13(self) -> None:
        path = test_data.get_path(
            "data-files/base/13/TIFF/current/n41w106/USGS_13_n41w106.xml"
        )
        item = stac.create_item(path)
        self.assertEqual(item.id, "n41w106-13")

    @pytest.mark.skip("FTP server doesn't resolve")
    def test_create_item_with_base(self) -> None:
        path = test_data.get_path(
            "data-files/base/1/TIFF/current/n41w106/USGS_1_n41w106.xml"
        )
        item = stac.create_item(path, base=USGS_FTP_BASE)
        data = item.assets["data"]
        self.assertEqual(
            data.href,
            (
                "ftp://rockyftp.cr.usgs.gov/vdelivery/Datasets/Staged"
                "/Elevation/1/TIFF/current/n41w106/USGS_1_n41w106.tif"
            ),
        )
        data = item.assets["metadata"]
        self.assertEqual(
            data.href,
            (
                "ftp://rockyftp.cr.usgs.gov/vdelivery/Datasets/Staged"
                "/Elevation/1/TIFF/current/n41w106/USGS_1_n41w106.xml"
            ),
        )
        data = item.assets["thumbnail"]
        self.assertEqual(
            data.href,
            (
                "ftp://rockyftp.cr.usgs.gov/vdelivery/Datasets/Staged/Elevation/"
                "1/TIFF/current/n41w106/USGS_1_n41w106.jpg"
            ),
        )
        link = next(link for link in item.links if link.rel == "via")
        self.assertTrue(link is not None)
        self.assertEqual(
            link.target,
            (
                "ftp://rockyftp.cr.usgs.gov/vdelivery/Datasets/Staged"
                "/Elevation/1/TIFF/current/n41w106/USGS_1_n41w106.xml"
            ),
        )

    @pytest.mark.skip("FTP server doesn't resolve")
    def test_create_item_from_product_and_id(self) -> None:
        path = test_data.get_path("data-files/base")
        item = stac.create_item_from_product_and_id("1", "n41w106", base=path)
        item.validate()

    def test_read_href_modifier(self) -> None:
        did_it = False

        def modify_href(href: str) -> str:
            nonlocal did_it
            did_it = True
            return href

        path = test_data.get_path(
            "data-files/base/1/TIFF/current/n41w106/USGS_1_n41w106.xml"
        )
        _ = stac.create_item(path, modify_href)
        self.assertTrue(did_it)


tests/test_utils.py
import unittest

from stactools.threedep import utils


class PathTest(unittest.TestCase):
    def test_simple(self) -> None:
        path = utils.path("1", "n41w106")
        self.assertEqual(path, "1/TIFF/current/n41w106/USGS_1_n41w106")

    def test_extension(self) -> None:
        path = utils.path("1", "n41w106", extension="foo")
        self.assertEqual(path, "1/TIFF/current/n41w106/USGS_1_n41w106.foo")

    def test_base(self) -> None:
        path = utils.path("1", "n41w106", base="foo")
        self.assertEqual(path, "foo/1/TIFF/current/n41w106/USGS_1_n41w106")


tests/test_commands.py
import os.path
from tempfile import TemporaryDirectory
from typing import Callable, List

import pystac
from click import Command, Group
from pystac import Catalog, Collection
from pystac.extensions.item_assets import ItemAssetsExtension
from stactools.testing.cli_test import CliTestCase

from stactools.threedep.commands import create_threedep_command


class CreateCollectionTest(CliTestCase):
    def create_subcommand_functions(self) -> List[Callable[[Group], Command]]:
        return [create_threedep_command]

    def test_create_collection(self) -> None:
        with TemporaryDirectory() as directory:
            result = self.run_command(
                f"threedep create-catalog {directory} --asset-id n41w106 "
                f"--asset-id n40w106 --quiet"
            )
            self.assertEqual(result.exit_code, 0, msg="\n{}".format(result.output))
            catalog = pystac.read_file(os.path.join(directory, "catalog.json"))
            assert isinstance(catalog, Catalog)
            item_ids = set([item.id for item in catalog.get_all_items()])
            self.assertEqual(
                item_ids, set(["n40w106-1", "n40w106-13", "n41w106-1", "n41w106-13"])
            )

            for child in catalog.get_children():
                assert isinstance(child, Collection)
                item_assets = ItemAssetsExtension.ext(child).item_assets

                data = item_assets["data"]
                self.assertIsNone(data.title)
                self.assertIsNone(data.description)
                self.assertEqual(data.media_type, pystac.MediaType.COG)
                assert data.roles
                self.assertListEqual(data.roles, ["data"])

                metadata = item_assets["metadata"]
                self.assertIsNone(metadata.title)
                self.assertIsNone(metadata.description)
                self.assertEqual(metadata.media_type, pystac.MediaType.XML)
                assert metadata.roles
                self.assertListEqual(metadata.roles, ["metadata"])

                thumbnail = item_assets["thumbnail"]
                self.assertIsNone(thumbnail.title)
                self.assertIsNone(thumbnail.description)
                self.assertEqual(thumbnail.media_type, pystac.MediaType.JPEG)
                assert thumbnail.roles
                self.assertListEqual(thumbnail.roles, ["thumbnail"])

                gpkg = item_assets["gpkg"]
                self.assertIsNone(gpkg.title)
                self.assertIsNone(gpkg.description)
                self.assertEqual(gpkg.media_type, pystac.MediaType.GEOPACKAGE)
                assert gpkg.roles
                self.assertListEqual(gpkg.roles, ["metadata"])


src/stactools/threedep/stac.py
import datetime
from typing import Optional

from pystac import Item
from pystac.extensions.raster import DataType, RasterBand, RasterExtension, Sampling
from pystac.extensions.scientific import ScientificExtension
from stactools.core import create
from stactools.core.io import ReadHrefModifier

from stactools.threedep.metadata import Metadata


def create_item(
    href: str,
    read_href_modifier: Optional[ReadHrefModifier] = None,
    base: Optional[str] = None,
) -> Item:
    """Creates a STAC item from an href to an XML metadata file."""
    metadata = Metadata.from_href(href, read_href_modifier)
    return create_item_from_metadata(metadata, read_href_modifier, base)


def create_item_from_product_and_id(
    product: str,
    id: str,
    read_href_modifier: Optional[ReadHrefModifier] = None,
    base: Optional[str] = None,
) -> Item:
    """Creates a STAC item from a product (e.g. "1") and an ID (e.g. "n41w106")."""
    metadata = Metadata.from_product_and_id(product, id)
    return create_item_from_metadata(metadata, read_href_modifier, base)


def create_item_from_metadata(
    metadata: Metadata,
    read_href_modifier: Optional[ReadHrefModifier] = None,
    base: Optional[str] = None,
) -> Item:
    """Creates a STAC item from Metadata."""
    href = metadata.asset_href_with_extension(base, "tif")
    item = create.item(href, read_href_modifier)
    item.id = metadata.stac_id
    end_datetime = metadata.end_datetime
    item.datetime = end_datetime
    item.common_metadata.start_datetime = metadata.start_datetime
    item.common_metadata.end_datetime = metadata.end_datetime
    item.links.append(metadata.via_link(base))

    scientific = ScientificExtension.ext(item, add_if_missing=True)
    assert metadata.publication_datetime
    scientific.citation = (
        f"U.S. Geological Survey, {metadata.publication_datetime.year}, {metadata.title}, "
        f"accessed {datetime.date.today().isoformat()}"
    )

    data = metadata.data_asset(base)
    item.add_asset("data", data)
    raster = RasterExtension.ext(data, add_if_missing=True)
    raster.bands = [
        RasterBand.create(
            nodata=-999999.0,
            sampling=Sampling.POINT,
            data_type=DataType.FLOAT32,
            spatial_resolution=metadata.spatial_resolution,
        )
    ]
    item.add_asset("metadata", metadata.metadata_asset(base))
    item.add_asset("thumbnail", metadata.thumbnail_asset(base))
    item.add_asset("gpkg", metadata.gpkg_asset(base))
    return item


src/stactools/threedep/utils.py
import os.path
from ftplib import FTP
from typing import List, Optional

import boto3

from stactools.threedep import utils
from stactools.threedep.constants import AWS_BUCKET, AWS_PREFIX, USGS_FTP_SERVER


def fetch_ids(product: str, use_usgs_ftp: bool = False) -> List[str]:
    """Returns all ids for the given product."""
    if use_usgs_ftp:
        return _fetch_ids_from_usgs_ftp(product)
    else:
        return _fetch_ids_from_aws(product)


def _fetch_ids_from_usgs_ftp(product: str) -> List[str]:
    directory = f"vdelivery/Datasets/Staged/Elevation/{product}/TIFF/current"
    ftp = FTP(USGS_FTP_SERVER)
    ftp.login()
    ids = [
        file_name
        for (file_name, _) in ftp.mlsd(directory)
        if not file_name.startswith(".")
    ]
    ftp.close()
    return ids


def _fetch_ids_from_aws(product: str) -> List[str]:
    path = os.path.dirname(utils.path(product, ""))
    prefix = os.path.join(AWS_PREFIX, path)
    client = boto3.client("s3")
    paginator = client.get_paginator("list_objects_v2")
    page_iterator = paginator.paginate(Bucket=AWS_BUCKET, Prefix=prefix)
    filtered_iterator = page_iterator.search("Contents[?ends_with(Key, `.xml`)].Key")
    # The main tif/xml files for each id are named simply like "USGS_1_n41w106.tif".
    # If there's been updates, older versions will have a datetime on the end, e.g.
    # "USGS_1_n41w106_20210330.tif" or something similar. By splitting on underscores,
    # we're hoping to catch only the "main" files.
    return [
        os.path.basename(os.path.dirname(key))
        for key in filtered_iterator
        if len(os.path.basename(key).split("_")) == 3
    ]


def path(
    product: str,
    id: str,
    base: Optional[str] = None,
    extension: Optional[str] = None,
    id_only: bool = False,
) -> str:
    """Returns the subpath for this product and id.

    E.g. path("1", "n41w106") == "1/TIFF/current/n41w106/USGS_1_n41w106"
    E.g. path("1", "n41w106", extension="tif") == "1/TIFF/current/n41w106/USGS_1_n41w106.tif"
    E.g. path("1", "n41w106", base="/base/dir") == "/base/dir/1/TIFF/current/n41w106/USGS_1_n41w106"
    E.g. path("1", "n41w106", id_only=True) == "/base/dir/1/TIFF/current/n41w106/n41w106"
    """
    if id_only:
        path = "{}/TIFF/current/{}/{}".format(product, id, id)
    else:
        path = "{}/TIFF/current/{}/USGS_{}_{}".format(product, id, product, id)
    if extension:
        path = "{}.{}".format(path, extension)
    if base:
        path = os.path.join(base, path)
    return path


src/stactools/threedep/metadata.py
import datetime
from typing import Optional, Union

from pystac import Asset, Link, MediaType
from stactools.core.io import ReadHrefModifier
from stactools.core.io.xml import XmlElement

from stactools.threedep import utils
from stactools.threedep.constants import DEFAULT_BASE


class Metadata:
    """3DEP file metadata."""

    @classmethod
    def from_href(
        cls, href: str, read_href_modifier: Optional[ReadHrefModifier] = None
    ) -> "Metadata":
        """Creates a metadata from an href to the XML metadata file."""
        xml_element = XmlElement.from_file(href, read_href_modifier)
        return cls(xml_element)

    @classmethod
    def from_product_and_id(
        cls, product: str, id: str, base: Optional[str] = None
    ) -> "Metadata":
        """Creates a Metadata from a product and id."""
        if base is None:
            base = DEFAULT_BASE
        href = utils.path(product, id, extension="xml", base=base)
        return cls.from_href(href)

    def __init__(self, xml: XmlElement):
        """Creates a new metadata object from XML metadata."""
        self.title = xml.find_text_or_throw(
            "./idinfo/citation/citeinfo/title", _missing_element
        )
        self.description = xml.find_text_or_throw(
            "./idinfo/descript/abstract", _missing_element
        )
        self.pubdate = xml.find_text_or_throw(
            "./idinfo/citation/citeinfo/pubdate", _missing_element
        )
        self.begdate = xml.find_text_or_throw(
            "./idinfo/timeperd/timeinfo/rngdates/begdate", _missing_element
        )
        self.enddate = xml.find_text_or_throw(
            "./idinfo/timeperd/timeinfo/rngdates/enddate", _missing_element
        )
        self.current = xml.find_text_or_throw(
            "./idinfo/timeperd/current", _missing_element
        )
        self.rowcount = xml.find_text_or_throw(
            "./spdoinfo/rastinfo/rowcount", _missing_element
        )
        self.colcount = xml.find_text_or_throw(
            "./spdoinfo/rastinfo/colcount", _missing_element
        )
        self.latres = xml.find_text_or_throw(
            "./spref/horizsys/geograph/latres", _missing_element
        )
        self.longres = xml.find_text_or_throw(
            "./spref/horizsys/geograph/longres", _missing_element
        )
        tiff_href = xml.find_text_or_throw(
            "./distinfo/stdorder/digform/digtopt/onlinopt/computer/networka/networkr",
            _missing_element,
        )
        parts = tiff_href.split("/")[-5:]
        # Some metadata have a 'current' or 'historical' in the path, some don't.
        if parts[2] == "TIFF":
            self.product = parts[1]
        else:
            self.product = parts[0]
        self.id = parts[3]

    @property
    def stac_id(self) -> str:
        """Returns the STAC ID of this metadata.

        This is the id plus the product, e.g. if the filename of the tif is
        "USGS_1_n40w105.tif", then the STAC id is "n40w105-1".
        """
        return "{}-{}".format(self.id, self.product)

    @property
    def publication_datetime(self) -> Optional[datetime.datetime]:
        """Returns the collection publication datetime."""
        if self.current == "publication date":
            return _format_date(self.pubdate)
        else:
            raise NotImplementedError

    @property
    def start_datetime(self) -> Optional[datetime.datetime]:
        """Returns the start datetime for this record.

        This can be a while ago, since the national elevation dataset was
        originally derived from direct survey data.
        """
        return _format_date(self.begdate)

    @property
    def end_datetime(self) -> Optional[datetime.datetime]:
        """Returns the end datetime for this record."""
        return _format_date(self.enddate, end_of_year=True)

    @property
    def spatial_resolution(self) -> float:
        """Returns the nominal ground sample distance from these metadata."""
        if self.product == "1":
            return 30
        elif self.product == "13":
            return 10
        else:
            raise NotImplementedError

    def data_asset(self, base: Optional[str] = DEFAULT_BASE) -> Asset:
        """Returns the data asset (aka the tiff file)."""
        return Asset(
            href=self.asset_href_with_extension(base, "tif"),
            title=self.title,
            description=None,
            media_type=MediaType.COG,
            roles=["data"],
        )

    def metadata_asset(self, base: Optional[str] = DEFAULT_BASE) -> Asset:
        """Returns the data asset (aka the tiff file)."""
        return Asset(
            href=self.asset_href_with_extension(base, "xml"),
            media_type=MediaType.XML,
            roles=["metadata"],
        )

    def thumbnail_asset(self, base: Optional[str] = DEFAULT_BASE) -> Asset:
        """Returns the thumbnail asset."""
        return Asset(
            href=self.asset_href_with_extension(base, "jpg"),
            media_type=MediaType.JPEG,
            roles=["thumbnail"],
        )

    def gpkg_asset(self, base: Optional[str] = DEFAULT_BASE) -> Asset:
        """Returns the geopackage asset."""
        description = (
            "Spatially-referenced polygonal footprints of the source data used "
            "to assemble the DEM layer. The attributes of each source dataset, "
            "such as original spatial resolution, production method, and date "
            "entered into the standard DEM, are linked to these footprints."
        )
        return Asset(
            href=self.asset_href_with_extension(base, "gpkg"),
            media_type=MediaType.GEOPACKAGE,
            roles=["metadata"],
            description=description,
        )

    def via_link(self, base: Optional[str] = DEFAULT_BASE) -> Link:
        """Returns the via link for this file."""
        return Link("via", self.asset_href_with_extension(base, "xml"))

    @property
    def region(self) -> str:
        """Returns this objects 3dep "region".

        Region is defined as a 10x10 lat/lon box that nominally contains this item.
        E.g. for n41w106, the region would be n40w110. This is used mostly for
        creating subcatalogs for STACBrowser.
        """
        import math

        n_or_s = self.id[0]
        lat = float(self.id[1:3])
        if n_or_s == "s":
            lat = -lat
        lat = math.floor(lat / 10) * 10
        e_or_w = self.id[3]
        lon = float(self.id[4:])
        if e_or_w == "w":
            lon = -lon
        lon = math.floor(lon / 10) * 10
        return f"{n_or_s}{abs(lat)}{e_or_w}{abs(lon)}"

    def asset_href_with_extension(
        self, base: Optional[str], extension: str, id_only: bool = False
    ) -> str:
        if base is None:
            base = DEFAULT_BASE
        return utils.path(
            self.product, self.id, base=base, extension=extension, id_only=id_only
        )


def _format_date(
    date: str, end_of_year: bool = False
) -> Union["datetime.datetime", None]:
    if len(date) == 4:
        year = int(date)
        if end_of_year:
            month = 12
            day = 31
        else:
            month = 1
            day = 1
        if year < 1800 or year > datetime.date.today().year:
            return None  # There's some bad metadata in the USGS records
        else:
            return datetime.datetime(
                year, month, day, 0, 0, 0, tzinfo=datetime.timezone.utc
            )
    elif len(date) == 8:
        year = int(date[0:4])
        month = int(date[4:6])
        day = int(date[6:8])
        return datetime.datetime(
            year, month, day, 0, 0, 0, tzinfo=datetime.timezone.utc
        )
    else:
        return None


def _missing_element(xpath: str) -> Exception:
    raise ValueError(f"missing required element: {xpath}")


src/stactools/threedep/constants.py
# flake8: noqa

from pyproj import CRS
from pystac import Provider, ProviderRole

USGS_3DEP_ID = "usgs-3dep"
THREEDEP_EPSG = 5498
THREEDEP_CRS = CRS.from_epsg(THREEDEP_EPSG)
PRODUCTS = ["1", "13"]
LICENSE = "PDDL-1.0"

DESCRIPTION = """The USGS 3D Elevation Program (3DEP) Datasets from The National Map are the primary elevation data product produced and distributed by the USGS. The 3DEP program provides a variety of resolution raster elevation data of the conterminous United States, Alaska, Hawaii, and the island territories. Some of the data sets such as the 1/3rd arc-second and 1 arc-second data set are derived from diverse source data sets that are processed to a specification with a consistent resolution, coordinate system, elevation units, and horizontal and vertical datums.  These seamless DEMs were referred to as the National Elevation Dataset (NED) from about 2000 through 2015 at which time they became the seamless DEM layers under the 3DEP program and the NED name and system were retired. Other 3DEP products include one-meter DEMs produced exclusively from high resolution light detection and ranging (lidar) source data and five-meter DEMs in Alaska as well as various source datasets including the lidar point cloud and interferometric synthetic aperture radar (Ifsar) digital surface models and intensity images. All 3DEP products are public domain. The 3DEP program is the logical result of the maturation of the long-standing USGS elevation program, which for many years concentrated on production of topographic map quadrangle-based digital elevation models. The 3DEP data  serves as the elevation layer of The National Map, and provides basic elevation information for earth science studies and mapping applications in the United States.

The seamless DEM layers under the 3DEP program are a multi-resolution dataset that is updated continuously to integrate newly available, improved elevation source data. Seamless DEM data layers under the 3DEP program  are available nationally at grid spacings of 1 arc-second (approximately 30 meters) for the conterminous United States, and at 1/3, 1/9 arc-seconds  (approximately 10 and 3 meters, respectively) and 1 meter for parts of the United States. Most seamless DEM data for Alaska is available at 2-arc-second (about 60 meters) grid spacing, where only lower resolution source data exist. Part of Alaska is available at the 1/3 arc-second, 1 arc-second and 5 meter resolution. Efforts are continuing to have full coverage of the 5 meter elevation data over Alaska in the next couple of years.
"""

USGS_PROVIDER = Provider(
    name="USGS",
    roles=[
        ProviderRole.PRODUCER,
        ProviderRole.PROCESSOR,
        ProviderRole.LICENSOR,
        ProviderRole.HOST,
    ],
    url="https://www.usgs.gov/core-science-systems/ngp/3dep",
)

USGS_FTP_SERVER = "rockyftp.cr.usgs.gov"
USGS_FTP_BASE = f"ftp://{USGS_FTP_SERVER}/vdelivery/Datasets/Staged/Elevation"
AWS_BUCKET = "prd-tnm"
AWS_PREFIX = "StagedProducts/Elevation"
AWS_BASE = f"https://{AWS_BUCKET}.s3.amazonaws.com/{AWS_PREFIX}"

DEFAULT_BASE = AWS_BASE


src/stactools/threedep/commands.py
import os.path
from collections import defaultdict
from typing import List

import click
from click import Command, Group
from pystac import Catalog, CatalogType, Collection, Extent, MediaType
from pystac.extensions.item_assets import AssetDefinition, ItemAssetsExtension
from pystac.extensions.raster import DataType, Sampling
from stactools.core.io import read_text

from stactools.threedep import stac, utils
from stactools.threedep.constants import (
    DESCRIPTION,
    PRODUCTS,
    USGS_3DEP_ID,
    USGS_FTP_BASE,
    USGS_PROVIDER,
)


def create_threedep_command(cli: Group) -> Command:
    """Creates the threedep command line utility."""

    @cli.group("threedep", short_help="Work with USGS 3DEP elevation data.")
    def threedep() -> None:
        pass

    @threedep.command(
        "create-catalog", short_help="Create a STAC catalog for existing USGS 3DEP data"
    )
    @click.argument("destination")
    @click.option(
        "-s",
        "--source",
        help="The href of a directory tree containing metadata",
        default=None,
    )
    @click.option(
        "-i",
        "--asset-id",
        "asset_ids",
        multiple=True,
        help="Asset ids to fetch. If not provided, will fetch all IDs.",
    )
    @click.option("--quiet/--no-quiet", default=False)
    @click.option("-c", "--catalog-type", type=str, default=CatalogType.SELF_CONTAINED)
    def create_catalog_command(
        destination: str,
        source: str,
        asset_ids: List[str],
        quiet: bool,
        catalog_type: CatalogType,
    ) -> None:
        """Creates a relative published 3DEP catalog in DESTINATION.

        If SOURCE is not provided, will use the metadata in AWS. SOURCE is
        expected to be a directory tree mirroring the structure on USGS, so
        it is best created using `stac threedep download-metadata`.
        """
        collections = {}
        items = defaultdict(list)
        for product in PRODUCTS:
            for asset_id in asset_ids or utils.fetch_ids(product):
                item = stac.create_item_from_product_and_id(
                    product, asset_id, base=source
                )
                items[product].append(item)
                if not quiet:
                    print(item.id)
            extent = Extent.from_items(items[product])
            if product == "1":
                title = "1 arc-second"
                description = "USGS 3DEP 1 arc-second DEMs"
                spatial_resolution = 30
            elif product == "13":
                title = "1/3 arc-second"
                description = "USGS 3DEP 1/3 arc-second DEMs"
                spatial_resolution = 10
            else:
                raise NotImplementedError
            collection = Collection(
                id=f"{USGS_3DEP_ID}-{product}",
                title=title,
                keywords=["USGS", "3DEP", "NED", "DEM", "elevation"],
                providers=[USGS_PROVIDER],
                description=description,
                extent=extent,
                license="PDDL-1.0",
            )
            item_assets = ItemAssetsExtension.ext(collection, add_if_missing=True)
            item_assets.item_assets = {
                "data": AssetDefinition(
                    {
                        "type": MediaType.COG,
                        "roles": ["data"],
                        "raster:bands": [
                            {
                                "nodata": -999999.0,
                                "sampling": Sampling.POINT,
                                "data_type": DataType.FLOAT32,
                                "spatial_resolution": spatial_resolution,
                            }
                        ],
                    }
                ),
                "metadata": AssetDefinition(
                    {"type": MediaType.XML, "roles": ["metadata"]}
                ),
                "thumbnail": AssetDefinition(
                    {"type": MediaType.JPEG, "roles": ["thumbnail"]}
                ),
                "gpkg": AssetDefinition(
                    {"type": MediaType.GEOPACKAGE, "roles": ["metadata"]}
                ),
            }
            collections[product] = collection
        catalog = Catalog(
            id=USGS_3DEP_ID,
            description=DESCRIPTION,
            title="USGS 3DEP DEMs",
            catalog_type=catalog_type,
        )
        for product, collection in collections.items():
            catalog.add_child(collection)
            collection.add_items(items[product])
        catalog.normalize_hrefs(destination)
        catalog.save()
        catalog.validate()

    @threedep.command(
        "download-metadata", short_help="Download all metadata for USGS 3DEP data"
    )
    @click.argument("destination")
    @click.option(
        "-i",
        "--asset-id",
        "asset_ids",
        multiple=True,
        help="Asset ids to fetch. If not provided, will fetch all IDs.",
    )
    @click.option("--quiet/--no-quiet", default=False)
    def download_metadata_command(
        destination: str, asset_ids: List[str], quiet: bool
    ) -> None:
        """Creates a 3DEP collection in DESTINATION."""
        for product in PRODUCTS:
            if not asset_ids:
                asset_ids = utils.fetch_ids(product)
            for id in asset_ids:
                path = utils.path(product, id, extension="xml", base=destination)
                if os.path.exists(path):
                    if not quiet:
                        print("{} exists, skipping download...".format(path))
                    continue
                os.makedirs(os.path.dirname(path), exist_ok=True)
                source_path = utils.path(
                    product, id, extension="xml", base=USGS_FTP_BASE
                )
                if not quiet:
                    print("{} -> {}".format(source_path, path))
                text = read_text(source_path)
                with open(path, "w") as f:
                    f.write(text)

    @threedep.command(
        "fetch-ids", short_help="Fetch all product ids and print them to stdout"
    )
    @click.argument("product")
    @click.option(
        "--usgs-ftp/--no-usgs-ftp",
        default=False,
        help="Fetch from the USGS FTP instead of AWS.",
    )
    def fetch_ids_command(product: str, usgs_ftp: bool) -> None:
        """Fetches product ids and prints them to stdout."""
        for id in utils.fetch_ids(product, use_usgs_ftp=usgs_ftp):
            print(id)

    return threedep
